### PR TITLE
C++2 Backend: Feature Tests Passing

### PR DESCRIPTION
**Pull request by sffc**
This PR adds struct and enum conversions. The feature test suite is passing now. Writeables/strings are still needed for the examples suite to pass.

### CODE REVIEW

**Review by Manishearth — COMMENTED** [review]
reached gen_method

**Review by Manishearth — APPROVED** [review]

File: tool/src/cpp2/ty.rs [hunk 1]
```diff
@@ -378,31 +520,99 @@ inline {ty_name}::~{ty_name}() {{
...
+                    Some(ReturnType::Writeable) => "/* TODO: Writeable conversion */".into(),
+                    None => "".into(),
+                    Some(ReturnType::OutType(o)) => self.gen_c_to_cpp(o, &ok_path),
+                };
+                let err_conversion = match err {
+                    Some(o) => self.gen_c_to_cpp(o, &err_path),
+                    None => "".into(),
+                };
+                Some(
+                    format!("{var_name}.is_ok ? diplomat::result<{ok_ty_name}, {err_ty_name}>(diplomat::Ok<{ok_ty_name}>({ok_conversion})) : diplomat::result<{ok_ty_name}, {err_ty_name}>(diplomat::Err<{err_ty_name}>({err_conversion}))").into()
```
[the remark this answers is not in the document]
Comment (sffc) [reply tool/src/cpp2/ty.rs:614]: (not actionable)

File: tool/src/cpp2/ty.rs [hunk 2]
```diff
@@ -378,31 +520,99 @@ inline {ty_name}::~{ty_name}() {{
...
+            Type::Opaque(ref op) => {
+                let op_id = op.tcx_id.into();
+                let ty_name = self.cx.formatter.fmt_type_name(op_id);
+                // TODO: Add imports?
+                format!("*{ty_name}::FromFFI({var_name})").into()
+            }
+            Type::Struct(ref st) => {
+                let id = P::id_for_path(&st);
+                let ty_name = self.cx.formatter.fmt_type_name(id);
+                // TODO: Add imports?
```
[the remark this answers is not in the document]
Comment (sffc) [reply tool/src/cpp2/ty.rs:553]: Updated comment

File: tool/src/cpp2/ty.rs [hunk 3]
```diff
@@ -322,39 +449,54 @@ inline {ty_name}::~{ty_name}() {{
         }
     }
 
-    fn gen_cpp_to_c_param<'a, P: TyPosition>(
+    fn gen_cpp_to_c<'a, P: TyPosition>(
```
[the remark this answers is not in the document]
Comment (sffc) [reply tool/src/cpp2/ty.rs:452]: Will document the functions in one big PR later

File: tool/src/cpp2/ty.rs [hunk 4]
```diff
@@ -378,31 +520,99 @@ inline {ty_name}::~{ty_name}() {{
...
+                let span = self
+                    .cx
+                    .formatter
+                    .fmt_borrowed_slice(&prim_name, b.mutability);
+                format!("{span}({var_name}_data, {var_name}_size)").into()
+            }
+        }
+    }
+
+    fn gen_fallible_c_to_cpp<'a>(
```
[the remark this answers is not in the document]
Comment (sffc) [reply tool/src/cpp2/ty.rs:579]: Will document the functions in one big PR later

File: tool/src/cpp2/ty.rs [hunk 5]
```diff
@@ -59,17 +62,81 @@ pub struct TyGenContext<'ccx, 'tcx, 'header> {
 impl<'ccx, 'tcx: 'ccx, 'header> TyGenContext<'ccx, 'tcx, 'header> {
     pub fn gen_enum_def(&mut self, ty: &'tcx hir::EnumDef, id: TypeId) {
         let ty_name = self.cx.formatter.fmt_type_name(id);
-        writeln!(self.decl_header, "enum struct {ty_name} {{").unwrap();
+        let ctype = self.cx.formatter.fmt_c_name(&ty_name);
+        self.decl_header
+            .includes
+            .insert(self.cx.formatter.fmt_c_decl_header_path(id));
+        write!(
+            self.decl_header,
```
[the remark this answers is not in the document]
Comment (sffc) [reply tool/src/cpp2/ty.rs:70]: Done

File: tool/src/cpp2/formatter.rs [hunk 6]
```diff
@@ -52,6 +52,14 @@ impl<'tcx> Cpp2Formatter<'tcx> {
     pub fn fmt_enum_variant(&self, variant: &'tcx hir::EnumVariant) -> Cow<'tcx, str> {
         variant.name.as_str().into()
     }
+    pub fn fmt_c_enum_variant<'a>(
+        &self,
+        ident: &'a str,
+        variant: &'tcx hir::EnumVariant,
+    ) -> Cow<'tcx, str> {
+        let variant_name = &variant.name;
+        format!("capi::{ident}_{variant_name}").into()
```
[the remark this answers is not in the document]
Comment (sffc) [reply tool/src/cpp2/formatter.rs:61]: Done

**Review by Manishearth — APPROVED** [review]

**Pull request merged**